db/database.go

```diff
@@ -2323,3 +2103,206 @@ func (dbc *DatabaseContext) AuthenticatorOptions() auth.AuthenticatorOptions {
 func (dbc *DatabaseContext) GetRequestPlusSequence() (uint64, error) {
 	return dbc.sequences.getSequence()
 }
+
+func (db *DatabaseContext) StartOnlineProcesses(ctx context.Context) error {
+
+	// Callback that is invoked whenever a set of channels is changed in the ChangeCache
+	notifyChange := func(changedChannels channels.Set) {
+		db.mutationListener.Notify(changedChannels)
+	}
+
+	// Initialize the ChangeCache.  Will be locked and unusable until .Start() is called (SG #3558)
+	if err := db.changeCache.Init(
+		ctx,
+		db,
+		db.channelCache,
+		notifyChange,
+		db.Options.CacheOptions,
+		db.MetadataKeys,
+	); err != nil {
+		base.DebugfCtx(ctx, base.KeyDCP, "Error initializing the change cache", err)
+		return err
+	}
+	db.mutationListener.OnChangeCallback = db.changeCache.DocChanged
+
+	if base.IsEnterpriseEdition() {
+		cfgSG, ok := db.CfgSG.(*base.CfgSG)
+		if !ok {
+			return fmt.Errorf("Could not cast %V to CfgSG", db.CfgSG)
+		}
+		db.changeCache.cfgEventCallback = cfgSG.FireEvent
+	}
+
+	importEnabled := db.UseXattrs() && db.autoImport
+	sgReplicateEnabled := db.Options.SGReplicateOptions.Enabled
+
+	// Initialize node heartbeater in EE mode if sg-replicate or import enabled on the node.  This node must start
+	// sending heartbeats before registering itself to the cfg, to avoid triggering immediate removal by other active nodes.
+	if base.IsEnterpriseEdition() && (importEnabled || sgReplicateEnabled) {
+		// Create heartbeater
+		heartbeaterPrefix := db.MetadataKeys.HeartbeaterPrefix(db.Options.GroupID)
+		heartbeater, err := base.NewCouchbaseHeartbeater(db.MetadataStore, heartbeaterPrefix, db.UUID)
+		if err != nil {
+			return pkgerrors.Wrapf(err, "Error starting heartbeater for bucket %s", base.MD(db.Bucket.GetName()).Redact())
+		}
+		err = heartbeater.StartSendingHeartbeats()
+		if err != nil {
```

Comment:
The old code would stop the heartbeater here if there was an error below this, in:

```
		cleanupFunctions = append(cleanupFunctions, func() {
			dbContext.Heartbeater.Stop()
		})
```